init/deviceref.go

```diff
@@ -86,24 +88,87 @@ func (d *deviceRef) resolveFromGptTable(devPath string, t []gptPart) *deviceRef
+			if err == nil {
+				if info.format == "luks" {
+					debug("The found root partition has a LUKS header.")
+					name, err := parseLuksParameters(info)
+					if err != nil {
+						debug("LUKS kernel parameters aren't valid. Substituting them with autogenerated parameters to unlock the found root partition in /dev/mapper/root.")
+						cmdline["rd.luks.name"] = info.uuid.toString() + "=root"
+						return &deviceRef{refName, "/dev/mapper/root"}
+					} else if name != "" {
+						debug("LUKS kernel parameters already specified. Unlocked root is expected on /dev/mapper/%s.", name)
+						return &deviceRef{refName, "/dev/mapper/" + name}
+					} else {
+						_, nameSet := cmdline["rd.luks.name"]
+						_, uuidSet := cmdline["rd.luks.uuid"]
+						if nameSet || uuidSet {
+							debug("Found root partition and LUKS kernel parameters don't match. Searching for other partitions...", rootDevName) // This edge case isn't handled in the specification. This should provide the most compatible behaviour.
+							continue
+						} else {
+							debug("No LUKS kernel parameters were provided. Generating parameters to unlock the found root partition in /dev/mapper/root.")
+							cmdline["rd.luks.name"] = info.uuid.toString() + "=root"
+							return &deviceRef{refName, "/dev/mapper/root"}
+						}
+					}
+				} else if info.isFs && info.format != "lvm" && info.format != "mdraid" {
+					debug("The found root partition has a regular (not encrypted) filesystem. Using %s as root.", rootDevName)
+					return &deviceRef{refName, rootDevName}
+				} else if info.isFs {
+					debug("The found root partition has an LVM physical volume or an mdraid block device. Those are incompatible with autodiscovery. Searching for other partitions...")
+					continue
+				}
+			}
+			debug("The found root partition has an incompatible file system or isn't formatted. Searching for other partitions...")
+		}
+	}
+	debug("No suitable root partitions were found on %s.", devPath) // We don't fail yet as this point may be reached in the unlikely but still valid situation of two ESPs with the same GUID.
+	return d
+}
+
 func parseDeviceRef(name, param string, enableAutodetect bool) (*deviceRef, error) {
 	if param == "" {
 		// try to auto-discover gpt partition https://www.freedesktop.org/wiki/Specifications/DiscoverablePartitionsSpec/
-		if autodiscoveryGuid, ok := autodiscoveryGptTypes[runtime.GOARCH]; enableAutodetect && ok {
-			debug("%s= param is not specified. Use GPT partition autodiscovery with guid type %s", name, autodiscoveryGuid)
-			gptType, err := parseUUID(autodiscoveryGuid)
+		if enableAutodetect && efiVarsAvailable {
+			debug("%s= param is not specified. Use GPT partition autodiscovery.", name)
+
+			_, data, err := attributes.ReadEfivarsWithGuid("LoaderDevicePartUUID", *util.StringToGUID("4a67b082-0a4c-41cf-b6c7-440b29bb8c4f"))
+			if err != nil {
+				return nil, err
+			}
+
+			espUuid, err := util.ParseUtf16Var(data)
```

Comment:
We already have a function to convert utf16 into a string - `fromUnicode16`. Let's use own implementation instead of this third party util function.